File: 1056.Capacity To Ship Packages Within D Days/solution.cpp
class Solution {
public:
    bool helper(vector<int>& weights, int cap, int days) {
        int need = 1;
        int total = 0;
        for(int i:weights) {
            if(i > cap) return false;
            if(total + i > cap) {
                total = i;
                need++;
                if(need > days) {
                    return false;
                }
            }else{
                total += i;
            }

        }
        return need <= days;
    }
    int shipWithinDays(vector<int>& weights, int days) {
        int l = weights[0];
        int r = 0;
        
        for(int i:weights) {
            r += i;
            if(i >= l) {
                l = i;
            }
        }
        int res = 0;

        while(l < r) {
            int mid = l + (r - l - 1) / 2;

            if(helper(weights, mid, days)) {
                r = mid;
            }else {
                l = mid + 1;
            }
        }
        return r;
    }
};

File: 1056.Capacity To Ship Packages Within D Days/solution.java
class Solution {
    public boolean canship(int[] weights, int days, int cap){
        int totaldays = 1;
        int load = 0;
        for(int i:weights){
            if(load+i>cap){
                totaldays++;
                load = i;
                if(totaldays>days) return false;
            }else{
                load += i;
            }
        }
        return true;
    }
    public int shipWithinDays(int[] weights, int days) {
        int left = 0;
        int right = 0;
        for(int i:weights){
            if(i>left){
                left = i;
            }
            right +=i;
        }
        int res = right;

        while(left<=right){
            int mid = left+(right-left)/2;

            if(canship(weights,days,mid)){
                res = mid;
                right = mid - 1;
            }else{
                left = mid + 1;
            }
        }
        return res;
    }
}